---
title: Genome After-Party planning
subtitle: This page gives an overview of the changes we're planning for the Genome After-Party pipelines
---

# Assembly download pipeline

## Now {#insdcdownload-now}

- (feature) Generate a mapping file between accession numbers, chromosome numbers, and sequence names
  - v2.1.0
- (bug) Generate a correct SAM header for the CRAM files
  - v2.1.0

# Read mapping

## Now {#readmapping-now}

- Release features already implemented on the development branch
  - As v1.4.0
  - Fix alignment commands and parameters. Use the same code as TreeVal &amp; co
  - Switch to HiFi-trimmer and provide the ability to _not_ filter (that's what I actually want in production)
  - Support for ULI
- (bug) Update CPU/memory settings
  - v1.4.0 is such a big change it's unlikely to be requesting the right resources at the first attempt
  - Necessary for production but lengthy to do, so not fitting v1.4.0
  - For v1.4.1

## Later {#readmapping-later}

- Support for PiMMS ?
  - For v1.5.0
- Support for RNA-Seq
  - For v1.6.0

## Long-term goal {#readmapping-longterm}

- Take the generation of contact maps from genomenote
  - Check if we could rather just run curationpretext

# Variant calling and composition

## Now {#variantcalling-now}

- (feature) add filtering options
  - For variantcomposition v0.2.0
- (feature) rearrange the outputs to be more practical
  - For variantcomposition v0.2.0
- (feature) remove features (except the stats) now in variantcomposition
  - For variantcalling v1.2.0
- (feature) make the merging of the input read alignment optional
  - For variantcalling v1.2.0

## Later {#variantcalling-later}

- (feature) Add structural variation detection
  - For variantcalling v1.3.0

# BlobToolKit

## Now {#blobtoolkit-now}

- Release features already implemented on the development branch
  - v0.10.0 or v1.0.0 if changes below are included
- (bug) Busco output tar files don't contain the sequences
  - Means all Buscos done with BlobToolKit >= v0.7.0, <= v0.9.0 will have to be rerun
    - All BlobToolKit runs with version < v0.7.0 have to be rerun anyway to enable the chromosome grids
  - Need to be included in the next production release, by v1.0.0
- (feature) parameter validation for Blastn and Busco
  - Requested by many external users of the pipeline
  - Only thing needed to call it v1.0.0

## Later {#blobtoolkit-later}

- (bug) Fix accepting existing Buscos as input (useful for large genomes when we run Busco outside of the pipeline)
  - Ideally v1.0.0 but could be deferred to v1.0.1

## Long-term goal {#blobtoolkit-longterm}

- Accept pre-computed analyses and make all the embedded analyses optional
  - fasta_windows
  - Read coverage

# Genomenote

## Now {#genomenote-now}

- Release features already implemented on the development branch
  - As v2.2.0

## Later {#genomenote-later}

- Support for combined maps ?
  - Need to check if we can use the curationpretext pipeline instead
  - For v2.3.0
- Smudgeplots ?
  - For v2.3.0
- Support for pre-computed analyses ? (Busco, contact maps, BlobToolKit)
  - For even later. TBD

## Long-term goal {#genomenote-longterm}

- Retire.
  - Move all the analyses to other pipelines
  - The Genome Note Platform will be providing the ability
    to fill a template in from Genome After-Party data

# Busco

## Later {#busco-later}

- Complete and release the pipeline v1.0.0
  - Regular Busco commands
  - Supports odb10 and odb12
  - Can find lineage automatically using the taxonomy
  - Make keeping the individual Fasta files an option
- Ancestral painting
  - For v1.1.0

## Long-term goal {#busco-longterm}

- Reinstate the Nextflow port of Busco (to run Busco on large genomes)

# Sequence composition

## Later {#sequencecomposition-later}

- (feature) Add simple repeat finders:
  - Low complexity repeats from `Dustmasker`.
  - Inverted repeats from `einverted`.
  - LTR retrotransposons from `LTRharvest` and `LTRdigest`.
  - Tandem repeats from `trf`.
  - Telomeric repeat annotation (tool to be confirmed).
  - Centromeric repeat annotation (tool to be confirmed).
- (feature) Add comprehensive repeat finders such as EarlGreyTE or EDTA2.
- (feature) Add TRASH and Pantera.
- (feature) Add stainedglass and/or ModDotPlot for self-alignment.
- (feature) Mappability tracks.

## Long-term goal {#sequencecomposition-longterm}

- Take these from genomenote
  - genomescope
  - smudgeplot
  - completeness and QV

# Ensembl download pipelines

## Later {#ensembldownloads-later}

- Support the new FTP / data release model of Ensembl
- The insdcdownload pipeline should:
  - download RefSeq annotation ?
- The ensemblgenedownload pipeline should
  - change the sequence names in the GFF file to match the accession numbers
  - compute the GFF stats and its BUSCO scores
  - compute gene density tracks
- The ensemblrepeatdownload pipeline should
  - compute repeat density tracks
  - download RepeatModeler models

## Long-term goal {#ensembldownloads-longterm}

- Rethink the naming / split of all download pipelines, e.g.:
  - combine the two Ensembl pipelines into one ?
  - pipeline to download from RefSeq (refseqdownload ?)
  - implement ENA support in insdcdownload ?
  - rename insdcdownload it ncbidownload
  - add GCA assembly support in nf-core/fetchngs ?

# All pipelines

All pipelines need a refresh to incorporate the [naming convention](/docs/usage/gap_conventions)
and make the input parameter names uniform.
Those changes will likely result in a _major_ release as the interface for using the pipeline changes.

## Now {#all-now}

- Avoid introducing new outputs that do not comply with the naming convention.
- Name new parameters in accordance with existing parameters across all Genome After-Party pipelines.
- If a _minor_ release is pending on the `dev` branch, rename the _newly-introduced_ input parameters
  and output files (if possible) to avoid having to do a major version bump soon after.
- If a _major_ release is pending on the `dev` branch, rename _all_ input parameters and output files
  _now_ to avoid having to do a major version bump soon after.

## Later {#all-later}

- For all other pipelines, plan a major version bump with all necessary changes, considering dependencies
  with other pipelines.
