---
title: Genome After-Party planning
subtitle: This page gives an overview of the changes we're planning for the Genome After-Party pipelines
---

# Assembly download pipeline

## Now {#insdcdownload-now}

### v3.0.0

- (feature) Generate a mapping file between accession numbers, chromosome numbers, and sequence names
- (feature) Generate a correct SAM header for the CRAM files

Both in review:

- <https://github.com/sanger-tol/insdcdownload/pull/47>
- <https://github.com/sanger-tol/insdcdownload/pull/48>

### v3.1.0

- Extra output file that lists the sequences in the correct order for contact maps
  - will be passed to `samtools faidx` when reordering is needed

## Later {#insdcdownload-later}

- Rename to "ncbidownload" and download RefSeq annotations too
- _or_ implement ENA backend support to make it truly "insdcdownload" and have a separate "refseqdownload"

# Read mapping

## Now {#readmapping-now}

> Versioning is uncertain as it depends on whether some changes happen or not, and whether they are breaking.

### 1.4.0 (TBC)

- (feature) Add an option to skip PacBio filtering/trimming
  - Assuming it's OK to use the untrimmed reads for variant-calling

### 1.5.0

- (feature) Chunked alignment for all data types, using the same commands as in the other sanger-tol pipelines
  - short reads – _done_
  - long reads – _in progress_
  - will need a review of the number of file conversions and the number of duplicated processes – _to do_
- (feature) PacBio pre-processing
  - optional HiFi-trimmer – _in progress_
  - optional ULI (single barcode) – _in progress_
- (feature) Better CRAM outputs – _in progress_

## Later {#readmapping-later}

### 1.6.0

- Generate a contact-map
  - Pretext format, with the annotation script run
  - Will need a command-line option to include/exclude multi-mapping reads

### 2.0.0

- Replace the "sample" samplesheet column with two: one for the specimen and one for the sequencing run.
- Support for multiple PacBio barcodes
- Support for PiMMS (PacBio)
- Implement ONT preprocessing if needed
- Update the input parameter names and the output file names as per our new [standards](#standards).

### 2.1.0

- Support for transcriptomic (Illumina RNA-seq, PacBio Iso-Seq / Kinnex)
  - May need specific stats on top of the basic `samtools stats`
- Update CPU/memory settings
  - impracticable to complete in time for the main release
- Support for UltimaGen data

# Variant calling and composition

## Now {#variantcalling-now}

### variantcomposition 0.2.0

- (feature) expose options for filtering the VCF
- (feature) rearrange the outputs to be more practical

### variantcalling 1.2.0

- (feature) remove features (except the stats) released in variantcomposition 0.2.0
- (feature) make the merging of the input read alignment optional
  - Will make it work with the new specimen/run_accession approach

### variantcalling 2.0.0 and variantcomposition 2.0.0

- Update the input parameter names and the output file names as per our new [standards](#standards).

## Later {#variantcalling-later}

### variantcalling 2.1.0

- (feature) Add structural variation detection ?

# BlobToolKit

## Now {#blobtoolkit-now}

### 0.10.0

- (bug) BUSCO output tar files don't contain the sequences
- (feature) parameter validation for Blastn and BUSCO
- (bug) Fix accepting existing BUSCOs as input (useful for large genomes when we run BUSCO outside of the pipeline)

Release under review: <https://github.com/sanger-tol/blobtoolkit/pull/217>

### 1.0.0

- May need more static images. Compare with the genomenote pipeline
- Work off specimen and run_accession like the other pipelines
- Update the input parameter names and the output file names as per our new [standards](#standards).

## Later {#blobtoolkit-later}

## Long-term goal {#blobtoolkit-longterm}

- Accept pre-computed analyses and make all the embedded analyses optional
  - fasta_windows
  - Read coverage

# Genome note

## Now {#genomenote-now}

### 2.3.0

- Sync blobtk plots with blobtoolkit

### 3.0.0

- Drop the HiGlass contact-map generation sub-workflow
- Drop the genome metadata sub-workflow
- Update the input parameter names and the output file names as per our new [standards](#standards).

## Later {#genomenote-later}

### 4.0.0

- Drop the Pretext contact-map generation sub-workflow, once implemented and released in readmapping

## Long-term goal {#genomenote-longterm}

- Offload functionality to other, more specific, pipelines.
- Turn into a meta-pipeline that can run the other pipelines itself or take pre-existing outputs.

# Sequence composition

## Later {#sequencecomposition-later}

> This is a long list of features and we will need to decide whether all
> should be in this one pipeline, or split across multiple pipelines.

- (feature) Add simple repeat finders:
  - Low complexity repeats from `Dustmasker`.
  - Inverted repeats from `einverted`.
  - LTR retrotransposons from `LTRharvest` and `LTRdigest`.
  - Tandem repeats from `trf`.
  - Telomeric repeat annotation (tool to be confirmed).
  - Centromeric repeat annotation (tool to be confirmed).
- (feature) Add comprehensive repeat finders such as EarlGreyTE or EDTA2.
- (feature) Add TRASH and Pantera.
- (feature) Add stainedglass and/or ModDotPlot for self-alignment.
- (feature) Mappability tracks.
- (feature) Compute gene annotation with tools such as BRAKER, EGAPx, or Helixer

## Long-term goal {#sequencecomposition-longterm}

- Take these from genomenote
  - genomescope
  - smudgeplot
  - completeness and QV

# Ensembl download pipelines

## Now {#ensembldownloads-now}

### 2.1

- ensemblgenedownload and ensemblrepeatdownload:
  - Support the new FTP / data release model of Ensembl
- ensemblgenedownload:
  - change the sequence names in the GFF file to match the accession numbers

## Later {#ensembldownloads-later}

### 2.2

- ensemblgenedownload:
  - compute the GFF stats and its BUSCO scores
  - compute gene density tracks
- ensemblrepeatdownload:
  - compute repeat density tracks
  - download RepeatModeler models

## Long-term goal {#ensembldownloads-longterm}

- Combine into a single "ensembldownload" pipeline that could be expanded to other Ensembl data ?

# Ideas for new pipelines

> Just ideas at this stage. All to be confirmed.

## BUSCO

Input:

- Assembly (Fasta)
- BUSCO lineage list
- Taxon name / ID (optional, but required for "all" lineage)
- Ancestral linkage groups mapping (optional)
- Chromosome list (optional, for ancestral painting)

Features:

- "all" special lineage resolves to all parent lineages
- "basal" special lineage resolves to all three basal lineages
- Run BUSCO on all selected lineages
- Tidy up the output directories
- Option to keep the individual Fasta files an option
- Ancestral painting
- Support for large genomes by splitting BUSCO as Nextflow jobs ?

## Genome statistics

Input:

- List of haploid assemblies
- k-mer library _or_ set of PacBio reads
- Option to skip BUSCO

Features:

- Run the genome_statistics sub-workflow on each assembly, using the other assemblies (concatenated) as the "alternate" assembly
- Combine the QV and completeness scores into a table

## GFF import

Input:

- GFF file
- Chromosome name mapping

Features:

- Rename the sequence names in the GFF file when needed
- Compute the GFF stats
- Generate CDS and Protein Fasta files
- Compute the BUSCO scores
- Compute gene density tracks

## Repeats import

Input:

- BED file (or whatever format is convenient / common for repeats)
- Chromosome name mapping

Features:

- Rename the sequence names in the BED file when needed
- Generate a masked Fasta file
- Generate masking statistics
- Compute repeat density tracks

# Pipeline design standards {#standards}

All pipelines need a refresh to incorporate the [naming convention](/docs/usage/gap_conventions)
and make the input parameter names uniform.
Those changes will likely result in a _major_ release as the interface for using the pipeline changes.

## Now {#all-now}

- Avoid introducing new outputs that do not comply with the naming convention.
- Name new parameters in accordance with existing parameters across all Genome After-Party pipelines.
- If a _minor_ release is pending on the `dev` branch, rename the _newly-introduced_ input parameters
  and output files (if possible) to avoid having to do a major version bump soon after.
- If a _major_ release is pending on the `dev` branch, rename _all_ input parameters and output files
  _now_ to avoid having to do a major version bump soon after.

## Later {#all-later}

- For all other pipelines, plan a major version bump with all necessary changes, considering dependencies
  with other pipelines.
